lntest/bitcoind.go

```diff
@@ -0,0 +1,184 @@
+// +build bitcoind
+
+package lntest
+
+import (
+	"fmt"
+	"io/ioutil"
+	"math/rand"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"time"
+
+	"github.com/btcsuite/btcd/chaincfg"
+	"github.com/btcsuite/btcd/rpcclient"
+)
+
+// logDir is the name of the temporary log directory.
+const logDir = "./.backendlogs"
+
+// BitcoindBackendConfig is an implementation of the BackendConfig interface
+// backed by a Bitcoind node.
+type BitcoindBackendConfig struct {
+	rpcHost      string
+	rpcUser      string
+	rpcPass      string
+	zmqBlockPath string
+	zmqTxPath    string
+	p2pPort      int
+	rpcClient    *rpcclient.Client
+
+	// minerAddr is the p2p address of the miner to connect to.
+	minerAddr string
+}
+
+// A compile time assertion to ensure BitcoindBackendConfig meets the
+// BackendConfig interface.
+var _ BackendConfig = (*BitcoindBackendConfig)(nil)
+
+// GenArgs returns the arguments needed to be passed to LND at startup for
+// using this node as a chain backend.
+func (b BitcoindBackendConfig) GenArgs() []string {
+	var args []string
+	args = append(args, "--bitcoin.node=bitcoind")
+	args = append(args, fmt.Sprintf("--bitcoind.rpchost=%v", b.rpcHost))
+	args = append(args, fmt.Sprintf("--bitcoind.rpcuser=%v", b.rpcUser))
+	args = append(args, fmt.Sprintf("--bitcoind.rpcpass=%v", b.rpcPass))
+	args = append(args, fmt.Sprintf("--bitcoind.zmqpubrawblock=%v",
+		b.zmqBlockPath))
+	args = append(args, fmt.Sprintf("--bitcoind.zmqpubrawtx=%v",
+		b.zmqTxPath))
+
+	return args
+}
+
+// ConnectMiner is called to establish a connection to the test miner.
+func (b BitcoindBackendConfig) ConnectMiner() error {
+	return b.rpcClient.AddNode(b.minerAddr, rpcclient.ANAdd)
+}
+
+// DisconnectMiner is called to disconnect the miner.
+func (b BitcoindBackendConfig) DisconnectMiner() error {
+	return b.rpcClient.AddNode(b.minerAddr, rpcclient.ANRemove)
+}
+
+// Name returns the name of the backend type.
+func (b BitcoindBackendConfig) Name() string {
+	return "bitcoind"
+}
+
+// NewBackend starts a bitcoind node and returns a BitoindBackendConfig for
+// that node.
+func NewBackend(miner string, netParams *chaincfg.Params) (
+	*BitcoindBackendConfig, func(), error) {
+
+	if netParams != &chaincfg.RegressionNetParams {
+		return nil, nil, fmt.Errorf("only regtest supported")
+	}
+
+	if err := os.MkdirAll(logDir, 0700); err != nil {
+		return nil, nil, err
+	}
+
+	logFile, err := filepath.Abs(logDir + "/bitcoind.log")
+	if err != nil {
+		return nil, nil, err
+	}
+
+	tempBitcoindDir, err := ioutil.TempDir("", "bitcoind")
+	if err != nil {
+		return nil, nil,
+			fmt.Errorf("unable to create temp directory: %v", err)
+	}
+
+	zmqBlockPath := "ipc:///" + tempBitcoindDir + "/blocks.socket"
+	zmqTxPath := "ipc:///" + tempBitcoindDir + "/txs.socket"
+	rpcPort := rand.Int()%(65536-1024) + 1024
+	p2pPort := rand.Int()%(65536-1024) + 1024
+
+	bitcoind := exec.Command(
+		"bitcoind",
+		"-datadir="+tempBitcoindDir,
+		"-regtest",
+		"-txindex",
```

Comment:
Should we run without this? Would ensure that the fallback path (no txindex, scan manually) is tested instead.

Reply:
We _could_ run it with both options, though this would significantly increase the running time of the current integration tests. On the other hand, we properly test both variants in the unit tests of the notifier.

Reply:
Could randomize it lol